src/ast.rs

```diff
@@ -0,0 +1,162 @@
+            })
+    }
+}
+
+impl<'s> Expression<Spanned<'s>> {
+    /// Parses the given `input` as an [`Expression`].
+    ///
+    /// # Errors
+    ///
+    /// See [`parse::Error`] for details.
+    pub fn parse<I: AsRef<str>>(
+        input: &'s I,
+    ) -> Result<Self, parse::Error<Spanned<'s>>> {
+        Self::try_from(input.as_ref())
+    }
+}
+
+/// Single entry of a [`cucumber-expression`][3].
+///
+/// See [`parse::single_expression()`] for the detailed grammar and examples.
+///
+/// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum SingleExpression<Input> {
+    /// [`alternation`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Alternation(Alternation<Input>),
+
+    /// [`optional`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Optional(Optional<Input>),
+
+    /// [`parameter`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Parameter(Parameter<Input>),
+
+    /// Text without whitespaces.
+    Text(Input),
+
+    /// Whitespaces are treated as a special case to avoid lookaheads and
+    /// lookbehinds described in the [architecture][1]. This allows parsing to
+    /// have `O(n)` complexity.
+    ///
+    /// [1]: https://tinyurl.com/cucumber-expr-spec
+    Whitespace,
+}
+
+/// Allows to match one of [`SingleAlternation`]s.
+///
+/// See [`parse::alternation()`] for detailed syntax and examples.
+#[derive(AsRef, Clone, Debug, Deref, DerefMut, Eq, PartialEq)]
+pub struct Alternation<Input>(pub Vec<SingleAlternation<Input>>);
+
+/// Building block an [`Alternation`].
+pub type SingleAlternation<Input> = Vec<Alternative<Input>>;
+
+impl<Input: InputLength> Alternation<Input> {
+    /// Returns length of capture from `Input`.
+    pub(crate) fn span_len(&self) -> usize {
+        self.0
+            .iter()
+            .flatten()
+            .map(|alt| match alt {
+                Alternative::Text(t) => t.input_len(),
+                Alternative::Optional(opt) => opt.input_len() + 2,
+            })
+            .sum::<usize>()
+            + self.len()
+            - 1
+    }
+
+    /// Indicates whether one of [`SingleAlternation`]s consists only from
+    /// [`Optional`]s.
+    pub(crate) fn contains_only_optional(&self) -> bool {
+        for single_alt in &**self {
+            if single_alt
+                .iter()
+                .all(|alt| matches!(alt, Alternative::Optional(_)))
+            {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// [`alternative`][3] expression.
+///
+/// See [`parse::alternative()`] for the detailed grammar and examples.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Alternative<Input> {
+    /// [`optional`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Optional(Optional<Input>),
+
+    /// Text.
+    Text(Input),
```

Comment:
This deviates from the [original grammar](https://github.com/cucumber/cucumber-expressions/blob/main/ARCHITECTURE.md#grammar) by missing the `parameter` variant.

Need to discuss this on voice.

Reply:
@tyranron there is a `Note:` section which says

> While `parameter` is allowed to appear as part of `alternative` and `option` in the AST, such an AST is not a valid a Cucumber Expression.

Basically `ARCHITECTURE.md` describes AST that tries to be wider than `Cucumber Expression` for some reason (the only I can think of is to make parser implementation simpler and then error on conversion to real `Cucumber Expression`). 
But in reality it doesn't make much sense to me. Especially `alternative` definition 
```ebnf
alternative         = optional | parameter | text
text                = whitespace | ")" | "}" | .
```
This grammar suggests that `alternative` may have unescaped whitespaces, which is not true: [example]. They [have to be escaped] at least to avoid ambiguity. 


[example]: https://cucumber.github.io/cucumber-expressions/?advanced=1&expression=alternation%2Fcan%27t%20have%20whitespaces&parameters=%5B%5D&step=alternation%20have%20whitespaces
[have to be escaped]: https://cucumber.github.io/cucumber-expressions/?advanced=1&expression=alternation%2Fcan%27t%5C%20have%20whitespaces&parameters=%5B%5D&step=can%27t%20have%20whitespaces

Reply:
It's not even EBNF, as I understand, as [wikipedia says] that repetition is described with `{...}` and not with `(...)*`. It looks more like regex. 



[wikipedia says]: https://en.wikipedia.org/wiki/Extended_Backus–Naur_form#Table_of_symbols

Reply:
@ilslv would you be so kind to make a PR to upstream that adjusts the described grammar to be accurate and precise enough. Because it really bothers: having spec which doesn't reflect reality, while implementations don't follow the spec 😕

Reply:
@ilslv 

> It's not even EBNF, as I understand, as [wikipedia says](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form#Table_of_symbols) that repetition is described with `{...}` and not with `(...)*`. It looks more like regex, but still has `,` for concatenation.

From your link: `*`  is a repetition, and `( ... )` is grouping. So we have a group repetion here. I don't see any mistakes in that. And in Markdown it has ` ```ebnf ` notation 🤷‍♂️